src/kubernetes.cc

```diff
@@ -982,20 +999,26 @@ void KubernetesUpdater::StartUpdater() {
     auto cb = [=](std::vector<MetadataUpdater::ResourceMetadata>&& results) {
       MetadataCallback(std::move(results));
     };
-    node_watch_thread_ = std::thread([=]() {
+    std::thread node_watch_thread = std::thread([=]() {
       reader_.WatchNodes(watched_node, cb);
     });
-    pod_watch_thread_ = std::thread([=]() {
+    object_watch_threads_.emplace(
+        WatchId("nodes", "v1"), std::move(node_watch_thread));
+    std::thread pod_watch_thread = std::thread([=]() {
       reader_.WatchPods(watched_node, cb);
     });
+    object_watch_threads_.emplace(
+        WatchId("pods", "v1"), std::move(pod_watch_thread));
     if (config().KubernetesClusterLevelMetadata() &&
         config().KubernetesServiceMetadata()) {
-      service_watch_thread_ = std::thread([=]() {
-        reader_.WatchObjects("services", "v1", cb);
-      });
-      endpoints_watch_thread_ = std::thread([=]() {
-        reader_.WatchObjects("endpoints", "v1", cb);
-      });
+      for (const auto& watch_id: watch_ids_) {
+        const std::string& plural_kind = watch_id.first;
```

Comment:
Careful — since the members of `WatchId` are now `const char*`, this becomes temporary, so we need to make it a value copy, rather than a reference. Same for `api_version`.